effekt/shared/src/main/scala/effekt/core/Deadcode.scala

```diff
@@ -41,69 +44,60 @@ object Deadcode {
  *
  * TODO this could also be extended to cover record and interface declarations.
  */
-class Reachable(
-  var reachable: Map[Id, Usage],
-  var stack: List[Id],
-  var seen: Set[Id]
-) {
-
-  def within(id: Id)(f: => Unit): Unit = {
-    stack = id :: stack
-    f
-    stack = stack.tail
+class Reachable(entrypoints: Set[Id], definitions: Map[Id, Definition]) {
+  private val reachable = mutable.Map[Id, Usage]()
+  private val seen = mutable.Set[Id]()
+  private val stack = mutable.Stack[Id]()
+
+  def analyze(): Map[Id, Usage] = {
+    entrypoints.foreach(process)
+    reachable.toMap
   }
 
-  def process(d: Definition)(using defs: Map[Id, Definition]): Unit =
-    if stack.contains(d.id) then
-      reachable = reachable.updated(d.id, Usage.Recursive)
-    else d match {
-      case Definition.Def(id, block) =>
-        seen = seen + id
-        within(id) { process(block) }
-
-      case Definition.Let(id, _, binding) =>
-        seen = seen + id
-        process(binding)
-    }
+  private def within(id: Id)(f: => Unit): Unit = {
+    stack.push(id)
+    f
+    stack.pop()
+  }
 
-  def process(id: Id)(using defs: Map[Id, Definition]): Unit =
+  private def process(id: Id): Unit = {
     if (stack.contains(id)) {
-      reachable = reachable.updated(id, Usage.Recursive)
-      return;
+      reachable(id) = Usage.Recursive
+      return
     }
 
-    val count = reachable.get(id) match {
-      case Some(Usage.Once) => Usage.Many
-      case Some(Usage.Many) => Usage.Many
-      case Some(Usage.Recursive) => Usage.Recursive
-      case None => Usage.Once
+    reachable.get(id) match {
+      case Some(Usage.Once) => reachable(id) = Usage.Many
+      case Some(Usage.Many | Usage.Recursive) => // Do nothing
+      case None => reachable(id) = Usage.Once
     }
-    reachable = reachable.updated(id, count)
+
     if (!seen.contains(id)) {
-      defs.get(id).foreach(process)
+      seen.add(id)
+      definitions.get(id).foreach(process)
     }
+  }
 
-  def process(b: Block)(using defs: Map[Id, Definition]): Unit =
-    b match {
-      case Block.BlockVar(id, annotatedTpe, annotatedCapt) => process(id)
-      case Block.BlockLit(tparams, cparams, vparams, bparams, body) => process(body)
-      case Block.Member(block, field, annotatedTpe) => process(block)
-      case Block.Unbox(pure) => process(pure)
-      case Block.New(impl) => process(impl)
-    }
+  private def process(d: Definition): Unit = d match {
+    case Definition.Def(id, block) =>
+      within(id) { process(block) }
+    case Definition.Let(id, _, binding) =>
+      process(binding)
+  }
+
+  @tailrec
+  private def process(b: Block): Unit = b match {
```

Comment:
This `private` is actually needed for `@tailrec`, the other ones are there for symmetry.
We could also mark it `final` instead.